Basic version of autoinstalling tasks

frontsideair opened the pull request:
We use libnpx to install tasks the same way npx autoinstalls packages to
a global cache, so the package is not installed globally. There's no
need for super user privileges this way.

Autoinstallation of presets is also changed to libnpx as well.

Fixes #97 

Any guidance for adding tests welcome!

Documentation changes may be necessary too.

sapegin reviewed (commented):
I think mocking `libnpx` is probably the only way to test it.

frontsideair commented on packages/mrm/package.json:
```diff
@@ -1,59 +1,61 @@
-    "tool",
-    "yaml"
-  ]
+	"name": "mrm",
```
Done!

frontsideair commented on packages/mrm/src/requirex.js:
```diff
@@ -0,0 +1,33 @@
+	return require.resolve(packagePath);
+};
+
+module.exports = requirex;
```
The original problem is that we may not have the given task or preset installed when the `mrm` command runs. The solution for presets was to interactively install that preset globally. But globally installing packages is not great for the reason I mentioned in the issue.

I found inspiration from the `npx` command which installs packages on-the-fly but without the downsides of globally installing them. I checked the source code and it seemed reusable. This is a tiny wrapper around `ensurePackages` from `libnpx` for this purpose.

sapegin replied:
This was autocomplete ;-( I was asking about the name — why `requirex`?

frontsideair replied:
Ah, because it uses `npx` under the hood, but I'm not too tied to the name. I'm also considering extracting it to a separate package but it's not a priority.

sapegin replied:
Shouldn't it be `requireUsingNpx` in this case? I think I'm still missing something ;-F

frontsideair replied:
I can rename, I was probably thinking of a cool package name. :D

frontsideair replied:
Done!

frontsideair commented on packages/mrm/src/requirex.js:
```diff
@@ -0,0 +1,33 @@
+const npx = require('libnpx');
+const path = require('path');
+
+async function requirex_(packageName) {
```
I hope my answer to the previous comment answers this as well, otherwise I can try to elaborate.

sapegin replied:
And here the question about composing file paths — it looks brittle.

frontsideair replied:
I was confused as the comment was pointing to the line above. I pulled this out of the [internal implementation](https://github.com/npm/npx/blob/latest/bin/index.js#L6) of `npx` with the only change being the addition of `..` as the script is under `src/`.

I'm not a fan either and open ot any suggestions.

sapegin replied:
I think we can use something like https://github.com/timoxley/npm-which since npm is a dependency.

frontsideair replied:
Maybe we can remove the `npm` dependency altogether, since `process.argv[0]` will always be `node` and it's always bundled with `npm`. I'll try this and if it doesn't work I'll give `npm-which` a shot.

frontsideair replied:
Ended up using `process.execPath` which is meant for this and is more robust. Worked like a charm!

frontsideair commented on packages/mrm/bin/mrm.js:
```diff
@@ -163,29 +156,11 @@ Note that when a preset is specified no default search locations are used.`
 			return [...paths, path.dirname(require.resolve('mrm-preset-default'))];
+		} else {
+			const presetPackageName = getPackageName('preset', preset);
+			const presetPath = await requirex.resolve(presetPackageName);
```
Note to self: This may fail if preset is local, add a test for this case.

frontsideair replied:
Added the test, but it doesn't run the `resolveDirectories` function, which I wanted to test. Still more tests is good.

frontsideair commented:
@sapegin I get a failed GitHub Actions email for every commit, is it broken or am I breaking things?

frontsideair commented on packages/mrm/src/index.js:
```diff
@@ -138,14 +137,23 @@ function getPackageName(type, packageName) {
+async function runTask(taskName, directories, options, argv) {
+	const taskPackageName = getPackageName('task', taskName);
+	const resolveList = [tryFile(directories, `${taskName}/index.js`)];
+	try {
```
This can be more elegant, I'll try to fix this before merging.

frontsideair replied:
Introduced `promiseFirst` for this.

frontsideair commented:
I think this is ready to go. I have written tests and updated internal messages to reflect that global installation is not necessary anymore. I didn't need to mock `libnpx` in the end, the tests run perfectly and give more confidence.

The only change I'm not sure about is where I promisified a few functions, I think it is for the better but I can easily revert it if it's not necessary.

I found only one reference to global installation in the docs and updated that.

Can you give it a look when you have time @sapegin?

sapegin reviewed (commented):
I think it looks much cleaner than the original implementation! I've left a few minor comments.

frontsideair commented on packages/mrm/package.json:
```diff
@@ -19,21 +19,19 @@
     "mrm-core": "^4.1.3",
     "mrm-preset-default": "^2.2.5",
-    "package-json": "6.5.0",
-    "requireg": "^0.2.2",
```
If I'm not mistaken, if the task or preset is not found in the given locations, it'll try to install with `npx`. So I guess the answer is no, but it won't break.

sapegin replied:
I think npx uses global or local packages if they are installed but it's worth checking, and likely documenting.

frontsideair replied:
`npx` does check local packages first (not sure about globals) but we're using the `_ensurePackages` function, which does not do that. I checked if that part could be simply added, but seems like there are hacks on top of hacks and too much code for little gain.

I can take another look but my impression is that people don't use globally installed packages anymore. For local packages we can first check `CWD/node_modules` first.

sapegin replied:
Then it's going to be a breaking change and a major release, which isn't a big problem.

The main problem I see is that npx is very very slow even if the package is cached already, so running an alias with a dozen of tasks could take a lot of time. If we can solve this problem, then it should be fine. I wanted to remove the default preset from the mrm package in the next major release anyway.

frontsideair replied:
Glad to hear that.

I'm thinking that I should be able to not resort to `npx` at least for locally installed packages, and will look into it now.

frontsideair replied:
The last commit I made makes this possible.

frontsideair replied:
I just noticed that this makes the behavior very weird, since we first check `taskPackageName` locally and then on npm registry, then we switch to `taskName`. The order should be:
- taskPackageName locally
- taskName locally
- taskPackageName on npm
- taskName on npm

I'll revert the last change and fix this.

frontsideair replied:
Should be fixed now.

frontsideair commented on packages/mrm/src/__tests__/index.spec.js:
```diff
@@ -64,110 +54,64 @@ const argv = {
-
-		return expect(installGlobalPackage('mrm-preset-default')).rejects.toThrow(
-			error
+		// ideally we can use toThrowError but that works with >= jest@22
```
I didn't check `package.json` but this was there for many other tests. I can do it for all of them or could be done in a separate PR.

sapegin replied:
A separate pull request is fine! Would be great!

frontsideair commented on packages/mrm/src/index.js:
```diff
@@ -320,13 +320,13 @@ function getConfig(directories, filename, argv) {
+async function getConfigFromFile(directories, filename) {
+	try {
+		const filepath = await tryFile(directories, filename);
+		return require(filepath);
```
Done!

frontsideair commented on packages/mrm/bin/mrm.js:
```diff
@@ -84,8 +83,8 @@ async function main() {
 							`“${taskName}” in the default mrm tasks`,
-							`npm install -g mrm-task-${taskName}`,
-							`npm install -g ${taskName}`,
+							`npm install mrm-task-${taskName}`,
```
How about this:
```suggestion
							`mrm-task-${taskName} package on the npm registry`,
```

frontsideair commented on packages/mrm/bin/mrm.js:
```diff
@@ -84,8 +83,8 @@ async function main() {
 							`“${taskName}” in the default mrm tasks`,
-							`npm install -g mrm-task-${taskName}`,
-							`npm install -g ${taskName}`,
+							`npm install mrm-task-${taskName}`,
+							`npm install ${taskName}`,
```
With the last change, my current suggestion is this:

```suggestion
							`mrm-task-${taskName} package in local node_modules`,
							`${taskName} package in local node_modules`,
							`mrm-task-${taskName} package on the npm registry`,
							`${taskName} package on the npm registry`,```

frontsideair commented:
I'm done for today, the last outstanding issue is the messaging and I have a suggestion for that. Feel free to apply it and merge if it looks okay! I can open a PR for Jest comments as soon as possible.

sapegin reviewed (approved):
Sorry for the delay! I think it looks good now, do you want to do anything else here, or it's good to merge?

We'll need to think how to properly release a new version (we should deprecate the default preset and remove it from the mrm package), and I'll have a look if we need to update any docs.

frontsideair commented:
Thanks for looking into this! I think this is ready, I just had a small improvement to the error message regarding search locations. Feel free to merge.

The next steps will be doing the Jest task (I can do it), deprecating the default preset and releasing a major version. I did touch documentation but it might be a good idea to see if it's good to go.

frontsideair commented:
Hi @sapegin,

I understand that you're offline currently, but you can let me know anytime that fits you if there's something I can do to move this further. I appreciate your efforts building and maintaining this project and it was a joy to contribute.

Thanks!

sapegin commented:
Hey, I've created an issue with all the things that I'd like to see in the new major release: #135, and I've created a new branch `next` for the release — we could merge this pull request there as soon as there are no conflicts. 

I'll appreciate any help with the rest of the items on the list.

frontsideair commented:
Thanks for the push. I'm currently on resolving conflicts and merge, and can clarify that this PR indeed autoinstalls presets as well.

frontsideair commented:
Conflicts are resolved, I was careful to use npm@7 to ensure v2 lockfiles. Feel free to merge anytime!
